iceoryx2-ffi/cxx/include/iox2/waitset.hpp

```diff
@@ -0,0 +1,222 @@
+    /// acquire the source.
+    /// If an interrupt- (`SIGINT`) or a termination-signal (`SIGTERM`) was received, it will exit
+    /// the loop and inform the user via [`WaitSetRunResult`].
+    auto run(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<WaitSetRunResult, WaitSetRunError>;
+
+    /// Tries to wait on the [`WaitSet`]. The provided callback is called for every attachment that
+    /// was triggered and the [`AttachmentId`] is provided as an input argument to acquire the
+    /// source.
+    /// If nothing was triggered the [`WaitSet`] returns immediately.
+    auto run_once(const iox::function<void(AttachmentId<S>)>& fn_call) -> iox::expected<void, WaitSetRunError>;
+
+    /// Returns the capacity of the [`WaitSet`]
+    auto capacity() const -> uint64_t;
+
+    /// Returns the number of attachments.
+    auto len() const -> uint64_t;
+
+    /// Returns true if the [`WaitSet`] has no attachments, otherwise false.
+    auto is_empty() const -> bool;
+
+    /// Attaches a [`Listener`] as notification to the [`WaitSet`]. Whenever an event is received on the
+    /// object the [`WaitSet`] informs the user in [`WaitSet::run()`] to handle the event.
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    ///
+    /// # Safety
+    ///
+    /// * The [`Listener`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_notification(const Listener<S>& listener) -> iox::expected<Guard<S>, WaitSetAttachmentError>;
+
+    /// Attaches a [`FileDescriptorView`] as notification to the [`WaitSet`]. Whenever an event is received on the
+    /// object the [`WaitSet`] informs the user in [`WaitSet::run()`] to handle the event.
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    ///
+    /// # Safety
+    ///
+    /// * The corresponding [`FileDescriptor`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_notification(FileDescriptorView file_descriptor) -> iox::expected<Guard<S>, WaitSetAttachmentError>;
+
+    /// Attaches a [`Listener`] as deadline to the [`WaitSet`]. Whenever the event is received or the
+    /// deadline is hit, the user is informed in [`WaitSet::run()`].
+    /// The object cannot be attached twice and the
+    /// [`WaitSet::capacity()`] is limited by the underlying implementation.
+    /// Whenever the object emits an event the deadline is reset by the [`WaitSet`].
+    ///
+    /// # Safety
+    ///
+    /// * The corresponding [`Listener`] must life at least as long as the returned [`Guard`].
+    /// * The [`Guard`] must life at least as long as the [`WaitsSet`].
+    auto attach_deadline(const Listener<S>& listener,
```

Comment:
I would again just call this `attach` and let the additional `deadline` argument tell the story.